### DISCUSSION TITLE
Search and replace functionality

### DISCUSSION THREAD

**Discussion by yarick2000** [directus/directus#6503] [Feature Requests]
Hello, I was thinking it will be nice to have search & replace functionality (with a preview option) in Admin UI. And not just within single collection, but globally, across all or just selected list of collections. So content editor making search first, getting list of results and then deciding per each specific finding if he is going to make replace for that or not. I guess it probably can be applied on selected types of fields, like numeric or textual ones and may be with only specific interfaces, like Text Input or Text Area, but still it can be super useful in cases when, for example there are lot articles, blogs, testimonials, etc. that referencing specific company/brand/product that changed it's name and so all relevant resources should be updated accordingly. Or there is a link to specific site that recently changed its domain and so again, all the materials that referencing that site/domain should be updated....

**Comment by benhaynes** [comment]
An interesting feature, but it seems very... dangerous! There are issues displaying content across Collections, so my first thought is that this should be _per collection_. Beyond that, we do have Bulk Edit... but that replaces the entire fieids value. Perhaps this could be a new option there for string replace. We'd also have to make sure there are efficient ways to handle this within all database vendors.

**Reply by rijkvanzanten** [reply]
> We'd also have to make sure there are efficient ways to handle this within all database vendors.

This is the main thing to me! MS SQL Server has a `REPLACE()` function we could rely on, but I have no idea about the other vendors

**Reply by yarick2000** [reply]
I think it pretty wide supported, probably by all SQL compatible engines. Regarding to display issues - I think there is no need to  preview all the value, but just part of it with highlighting the text that was found inside of that part. So for example you was looking for the 'dog' word and the field value is:
_He held out his hand for the dog to sniff, but Nikolai nuzzled his nose into it instead._
But the preview may contain just part of the sentence, something like '... for the **dog** to sniff, but...' and if user will need to see the full value then there will be link that will redirect him to the item details page.
Actually, what I was thinking, the preview may be represented as a grid view with the following columns:
a) Collection Name - the name of collection where the search keyword has been found
b) Field Name - the name of the field where  the search keyword has been found
c) Before Replace - the preview of the part of the field with highlighted text that user was searching for
d) After Replace - the preview of the part of the field with replaced text that will show how text/value will looks like after replacement will be done
e) Checkbox column (probably should be first by default) - can be used by user for selecting only specific found instances for which ones he would like to process replacement. By default everything is selected
f) Primary Key - the value of the primary key for the item where the search keyword has been found
One of the columns or may be the full row can be a link that will point to specific item details page (I guess it should be open in a separate window/tab) where user can check the full value.
Also there will be buttons Proceed and Cancel (at the top or bottom of the screen) for processing replacement for selected items and for cancelling it. Once replacement will be done user will get notification about how many items were processed and/or how many replacements were done.

**Reply by philleepflorence** [reply]
> Implemented something similar in D8 ([Search](https://github.com/philleepflorence/directus-8-extensions/blob/master/helpers/search.php) and [Database](https://github.com/philleepflorence/directus-8-extensions/blob/master/helpers/database.php)) Extensions but ended up permitting replace mainly in dev or staging environments due to roles, accountability, activity, and revisions. 

Combining App+API, what do you think about: 
- extending [search component](https://github.com/directus/directus/tree/main/app/src/views/private/components/search-input) to include permitted collections (if the conflict with roles is minimal) + 
- the [Table Layout](https://docs.directus.io/concepts/layouts/#table-layout) + 
- extending [Multiple Items Updates](https://docs.directus.io/reference/api/items/#update-multiple-items) to allow substring replace + 
- [Knex Modify](http://knexjs.org/#Builder-modify) + 
- RDMS SQL Replace Function?

**Reply by yarick2000** [reply]
I think we can use the combination of the following:
For search and replace within collection interface (the one that is opened by using /collections/{collection} route in the App) the **new control** (icon) can be added beside the search one. Once clicked, it will be expanded same way as the search control, but it will show a bit more elements inside, in addition to Search input field there will be also Replace for replace value and Fields for selecting fields that will be participated in search & replaced processing. There also will be Preview button that by default will be disabled, and so once user will finish to enter search string, replace value and will select fields, this button will became enabled. After that,  when user will click on that button,  there will be a special tabular layout shown with the fields that I've described in previous comment, except of the collection name (since we already within specific collection scope).  This layout, in addition to grid itself will have a button called Proceed  or Replace that is disabled by default, but will became enabled once user will select at least one of the records in the grid. Once he will click on that button the replacement will happen (I think **Multiple Items Updates** is most suitable here) and user will get indication how many search instances were replaced (or/and how many items in collection were updated). After that user will be redirected to a default view for collection.

**Reply by yarick2000** [reply]
For search and replace across multiple collections another new control can be added as part of collections interface (the one that is opened by using /collections route in the App.  It can be integrated similar way as for single collection, but with some differences.
Once user will click on the search & replace icon on top of the screen the following will happen:
a)  Search icon will be expanded and 2 input field will appear, one for search keyword and another one for replace value
b) 2 new columns will appear in the list - checkboxes column (selection column) on the left side and editable Fields column on the right. First column will allow user to select collections where he wants to make replacement and fields will allow to select fields that he want's to use for search for that specific collection
c) The button Preview will appear as well, and once user will finish to enter all the data and clicks on it there will be the same interface shown, as from the previous scenario for single collection. The only difference is that there will be additional column shown, called Collection Name, with the name of the collection where the search keyword has been found.  All other functionality will be then the same as for single collection... However, I guess replacement may take longer time in that case and may be that should be implemented as some kind of async job, not sure though where to show it's status... May be we can have a special page that will show status of all running jobs....

**Reply by yarick2000** [reply]
![image](https://user-images.githubusercontent.com/13006578/123659483-bd274880-d800-11eb-9023-9ffd1a7dd5a8.png)

**Comment by licitdev** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
